src/bpftrace.cpp

```diff
@@ -1313,37 +1335,84 @@ int BPFtrace::setup_perf_events()
   return 0;
 }
 
-void BPFtrace::poll_perf_events(bool drain)
+int BPFtrace::setup_ringbuf()
 {
-  const int timeout_ms = 100;
+  ringbuf_ = static_cast<struct ring_buffer *>(bpf_new_ringbuf(
+      maps[MapManager::Type::Ringbuf].value()->mapfd_, ringbuf_printer, this));
+  if (reset_rb_loss_counter())
+  {
+    LOG(ERROR) << "fail to init ringbuf loss counter";
+    return -1;
+  }
+  return 0;
+}
+
+void BPFtrace::teardown_output()
+{
+  if (feature_->has_map_ringbuf())
+    bpf_free_ringbuf(ringbuf_);
+
+  if (!feature_->has_map_ringbuf() || resources.needs_perf_event_map)
+    // Calls perf_reader_free() on all open perf buffers.
+    open_perf_buffers_.clear();
+}
+
+void BPFtrace::poll_output(bool drain)
+{
+  int ready;
+  bool do_poll_perf_event = !feature_->has_map_ringbuf() ||
+                            resources.needs_perf_event_map;
+  bool do_poll_ringbuf = feature_->has_map_ringbuf();
 
-  if (epollfd_ < 0)
+  if (do_poll_perf_event && epollfd_ < 0)
   {
     LOG(ERROR) << "Invalid epollfd " << epollfd_;
     return;
   }
 
-  auto events = std::vector<struct epoll_event>(online_cpus_);
   while (true)
   {
-    int ready = epoll_wait(epollfd_, events.data(), online_cpus_, timeout_ms);
-    if (ready < 0 && errno == EINTR && !BPFtrace::exitsig_recv) {
-      // We received an interrupt not caused by SIGINT, skip and run again
-      continue;
+    if (do_poll_perf_event)
+    {
+      ready = poll_perf_events();
+      if (ready < 0 && errno == EINTR && !BPFtrace::exitsig_recv)
+      {
+        // We received an interrupt not caused by SIGINT, skip and run again
+        if (!do_poll_ringbuf)
+          continue;
+      }
+      if (ready < 0 || (ready == 0 && (drain || finalize_)))
+      // Stop if either
+      //   * epoll_wait has encountered an error (eg signal delivery)
+      //   * There's no events left and we've been instructed to drain or
+      //     finalization has been requested through exit() builtin.
+      {
+        do_poll_perf_event = false;
+      }
     }
 
-    // Return if either
-    //   * epoll_wait has encountered an error (eg signal delivery)
-    //   * There's no events left and we've been instructed to drain or
-    //     finalization has been requested through exit() builtin.
-    if (ready < 0 || (ready == 0 && (drain || finalize_)))
+    if (do_poll_ringbuf)
     {
-      return;
+      // print loss events
+      read_rb_loss_counter();
+      ready = bpf_poll_ringbuf(ringbuf_, timeout_ms);
+      if (ready < 0 && !BPFtrace::exitsig_recv)
+      {
+        // We received an interrupt not caused by SIGINT, skip and run again
+        continue;
+      }
+      if (BPFtrace::exitsig_recv || (ready == 0 && (drain || finalize_)))
+      {
+        // Stop if either
+        //   * exit signal is received
+        //   * There's no events left and we've been instructed to drain or
+        //     finalization has been requested through exit() builtin.
+        do_poll_ringbuf = false;
+      }
```

Comment:
This is duplicated from above. Any way to avoid duplication? Maybe put the each conditional into a local variable in the loop body?